src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,94 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+
+    private long maxId;
+    private List<Product> products;
+
+    public ArrayListProductDao() {
+        this.products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public synchronized Product getProduct(Long id) throws NoSuchElementException {
+        return products.stream()
+                .filter(it -> id.equals(it.getId()))
+                .findAny()
+                .get();
     }
 
     @Override
-    public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+    public synchronized List<Product> findProducts() {
+        return products.stream()
+                .filter(it -> it.getPrice() != null)
+                .filter(it -> it.getStock() > 0)
+                .collect(Collectors.toList());
     }
 
     @Override
-    public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+    public synchronized void save(Product product) {
+
+        if (product.getId() == null) {
+            addProduct(product);
+
+        } else {
+
+            Product existedProductWithThisId = products.stream()
+                    .filter(it -> product.getId().equals(it.getId()))
+                    .findAny()
+                    .orElse(null);
+
+            if (existedProductWithThisId == null) {
+                addProduct(product);
+
+            } else {
+                existedProductWithThisId.setCode(product.getCode());
+                existedProductWithThisId.setCurrency(product.getCurrency());
+                existedProductWithThisId.setDescription(product.getDescription());
+                existedProductWithThisId.setImageUrl(product.getImageUrl());
+                existedProductWithThisId.setPrice(product.getPrice());
+                existedProductWithThisId.setStock(product.getStock());
```

Comment:
Can we make code a bit more readable by moving this to a separate method like populateExitingProduct()?

Reply:
done